### PR TITLE
[14.0] Add module edi_bank_statement_import

### PR DESCRIPTION
**Pull request by TeoGoddet**
This module plugs edi framework into account_statement_import from OCA/bank-statement-import
It allow to parse data retrieved trough the framework as bank statement.

It's heavily based on https://github.com/Noviat/account_ebics/blob/14.0/account_ebics/models/ebics_file.py and #451

### CODE REVIEW

File: edi_bank_statement_import/components/process.py
Comment (TeoGoddet) [inline edi_bank_statement_import/components/process.py]: Hi !
Here it's possible to have mutiple statements in a single xml file, that mean multiple records for one exchange.
I suppose that it's possible to create child exchange but i could not find infos/example on how it should be done.

Comment (simahawk) [reply edi_bank_statement_import/components/process.py]: @TeoGoddet we've discussed this recently. IMO a good approach could be:

1. make exchange_record.exchange_file computed and write-able
2. if the record is a child and there's no value look at the parent

If you take action on this, please do it in a separated PR.

File: edi_bank_statement_import/components/process.py
Comment (simahawk) [inline edi_bank_statement_import/components/process.py]: pls split this `process` method into smaller methods

File: edi_bank_statement_import/readme/DESCRIPTION.rst
Comment (simahawk) [inline edi_bank_statement_import/readme/DESCRIPTION.rst:7]: yes pls :)

File: edi_bank_statement_import/tests/edi_bank_statement_import_test.py
Comment (simahawk) [inline edi_bank_statement_import/tests/edi_bank_statement_import_test.py]: this code runs for every test. Please use `SavepointCase` + `setUpClass` and if you don't rely on tracking also use this line:
```
cls.env = cls.env(context=dict(cls.env.context, tracking_disable=True))
```
You'll see this approach in edi_oca too.

File: edi_bank_statement_import/tests/edi_bank_statement_import_test.py
Comment (simahawk) [inline edi_bank_statement_import/tests/edi_bank_statement_import_test.py]: here you are testing the base wizard, this should be done in the base module not here unless you are testing a special behavior provided by this new module.

File: edi_bank_statement_import/tests/edi_bank_statement_import_test.py
Comment (simahawk) [inline edi_bank_statement_import/tests/edi_bank_statement_import_test.py]: same.

File: edi_bank_statement_import/tests/edi_bank_statement_import_test.py
Comment (simahawk) [inline edi_bank_statement_import/tests/edi_bank_statement_import_test.py]: pls have a look at how tests are done in edi_oca and test the component.

**Review by simahawk — CHANGES_REQUESTED** [review]

File: edi_bank_statement_import/components/process.py
Comment (TeoGoddet) [reply edi_bank_statement_import/components/process.py]: done

File: edi_bank_statement_import/components/process.py
Comment (TeoGoddet) [reply edi_bank_statement_import/components/process.py]: #542

File: edi_bank_statement_import/tests/edi_bank_statement_import_test.py
Comment (TeoGoddet) [reply edi_bank_statement_import/tests/edi_bank_statement_import_test.py]: removed the tests

File: edi_bank_statement_import/tests/edi_bank_statement_import_test.py
Comment (TeoGoddet) [reply edi_bank_statement_import/tests/edi_bank_statement_import_test.py]: removed the tests

File: edi_bank_statement_import/tests/edi_bank_statement_import_test.py
Comment (TeoGoddet) [reply edi_bank_statement_import/tests/edi_bank_statement_import_test.py]: removed the tests

File: edi_bank_statement_import/tests/edi_bank_statement_import_test.py
Comment (TeoGoddet) [reply edi_bank_statement_import/tests/edi_bank_statement_import_test.py]: removed the tests

File: edi_bank_statement_import/readme/DESCRIPTION.rst
Comment (TeoGoddet) [reply edi_bank_statement_import/readme/DESCRIPTION.rst:7]: I'm not really able to, but let the message (inspired from #451)

**Comment by TeoGoddet** [comment]
Hello @OCA/edi-maintainers, 
I'm sorry to bother you, but I don't understand why the test fails on another module. Does someone have an idea ?

File: edi_bank_statement_import/readme/DESCRIPTION.rst
Comment (TeoGoddet) [reply edi_bank_statement_import/readme/DESCRIPTION.rst:7]: Done, please check it

**Comment by TeoGoddet** [comment]
Hello @simahawk, 
I've added a demo and some tests (based on your work for sale order [75b81519f61aa385c3b412be4049aeeaf60736b1](https://github.com/OCA/edi/commit/75b81519f61aa385c3b412be4049aeeaf60736b1))

**Review by etobella — APPROVED** [review]
Makes sense to me.

**Comment by etobella** [comment]
@simahawk As original reviewer, can you give us your opinions ? :pray:

File: edi_bank_statement_import/components/process.py
Comment (simahawk) [inline edi_bank_statement_import/components/process.py:66]: Not a blocker, but nice to have:

```suggestion
        file_data = base64.b64decode(wiz.statement_file)
        wiz.import_single_file(file_data, result)
        
        if not statement_ids:
            return self._handle_already_imported()
        if len(statement_ids) == 1:
            return self._handle_single_result(self.exchange_record, statement_ids[0])
        return self._handle_result(result["statement_ids"], wiz.statement_filename)
   
    def _handle_already_imported(self, wiz):
            self.exchange_record.sudo().message_post(
                body=_(
                    "You have already imported this file (%s) or this file "
                    "only contains already imported transactions."
                )
                % wiz.statement_filename
            )
            return False
            
    def _handle_result(self, statement_ids, filename):
            for statement_id in statement_ids:

                child_exchange = self.exchange_record.create(
                    {
                        "identifier": "%s_%s"
                        % (self.exchange_record.identifier, statement_id),
                        "type_id": self.exchange_record.type_id.id,
                        "backend_id": self.exchange_record.backend_id.id,
                        "edi_exchange_state": "input_processed",
                        "parent_id": self.exchange_record.id,
                    }
                )
                self._handle_single_result(child_exchange, statement_id)
                self.exchange_record.sudo().message_post(
                    body=_("Multiple Bank Statements created (See child exchanges)")
                )
            return True
```

Comment (simahawk) [reply edi_bank_statement_import/components/process.py:66]: Well, the formatting is a bit messed up, but I think you get my point ;)

File: edi_bank_statement_import/components/process.py
Comment (simahawk) [inline edi_bank_statement_import/components/process.py:87]: Recommendation for all the cases where you use `with_context`: never override the full ctx.
Always explode the keys to not override the rest:
```suggestion
        wiz = self.env[wiz_model].with_context(**wiz_ctx).sudo().create(wiz_vals)
```

File: edi_bank_statement_import/data/data.xml
Comment (simahawk) [inline edi_bank_statement_import/data/data.xml:2]: Do you really think we need a specific backend type for this?
We should probably have a generic one into a new module `edi_bank`?
In any case, if you keep it, please move it to its own file and make it noupdate.

File: edi_bank_statement_import/data/data.xml
Comment (simahawk) [inline edi_bank_statement_import/data/data.xml:7]: Creating an exchange type by default w/out noupdate is dangerous.

BTW personally I'm thinking about removing default exchange types from the modules I maintain because it can be dangerous too due to their high level of configuration.

I tend to say that you should leave this to the specific setup of the instance and move this to demo data.
Just a recommendation tho. I'd like to hear your feedback and @etobella 's too.

File: edi_bank_statement_import/readme/DESCRIPTION.rst
Comment (simahawk) [reply edi_bank_statement_import/readme/DESCRIPTION.rst:7]: See my comment above

File: edi_bank_statement_import/data/data.xml
Comment (simahawk) [inline edi_bank_statement_import/data/data.xml:5]:
```suggestion
        <field name="name">Bank Statement</field>
        <field name="code">bank_statement</field>
```

**Review by simahawk — CHANGES_REQUESTED** [review]
Thanks for your work, I left some more remarks.

**Pull request closed without merging**
